### PR TITLE
[SPARK-10731] [PYSPARK] [SQL] use 1 partition in take() of Python DataFrame

### PR DESCRIPTION
**Pull request by davies**
And optimize this special case (zero or one partition) in Limit, remove the unnecessary shuffle.

### CODE REVIEW

**Comment by davies** [comment]
cc @yhuai

**Comment by rxin** [comment]
Can you update the title to put the correct JIRA ticket? SPARK-10731

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -300,7 +300,7 @@ def take(self, num):
         >>> df.take(2)
         [Row(age=2, name=u'Alice'), Row(age=5, name=u'Bob')]
         """
-        return self.limit(num).collect()
+        return self.coalesce(1).limit(num).collect()
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: Can this be made more of a wrapper around the Scala stuff rather than doing its own coalescing?

Comment (davies) [inline python/pyspark/sql/dataframe.py]: How?

Comment (rxin) [inline python/pyspark/sql/dataframe.py]: Is it possible to just call scala DataFrame.take and get the result? You lose the socket thing if the number of rows is enormous, but that doesn't seem like a big deal for take.

Note that this changes the behavior of take in Python vs Scala. In Scala, you can still get parallelism, whereas in Python you coalesced it into a single partition, and as a result the degree of parallelism is now just 1.

**Pull request closed without merging**